### ISSUE TITLE
Destructive operations

### ISSUE THREAD

**Issue by stevelinton** [gap-system/gap#306]
This replaces issue #303 with a more general one.

The documentation of the Destructive matrix operations, of which there are quite a few:

    [1] Reference: DeterminantMatDestructive
    [2] Reference: TransposedMatDestructive
    [3] Reference: NullspaceMatDestructive
    [4] Reference: TriangulizedNullspaceMatDestructive
    [5] Reference: SolutionMatDestructive
    [6] Reference: ElementaryDivisorsMatDestructive
    [7] Reference: ElementaryDivisorsTransformationsMatDestructive
    [8] Reference: SemiEchelonMatDestructive
    [9] Reference: SemiEchelonMatsDestructive
    [10] Reference: BaseMatDestructive

varies in small ways.  Sometimes it says it "may" destroy the argument, sometimes it says it "will", sometimes that it "may (and probably will)". In most but not all cases it says that the matrix must be mutable, although in one or two cases IsMutable is missing from the declaration.  In one case TransposedMat it says " In all other cases the transposed is computed by <Ref Attr="TransposedMat"/>.", implying that it should work for immutable matrices.

It would be nice to tidy this up.  I propose that the documentation should always say that it "may" destroy the argument, and (where relevant) that the returned value "may" be identical to the argument and that it always require a mutable argument. 

In the rare case of mutable matrices where the destructive operation cannot be performed (I think that might just be transpose of non-square compressed matrices) it should just use the non-destructive version, which can be implemented by a slightly ranked down fallback method.

**Comment by frankluebeck** [comment]
Note that IsMutable is not the correct condition, the rows of the matrix must also be mutable and not in IsLockedRepresentationVector.

```
gap> mat:=[[1,2,3],[4,5,6]];
[ [ 1, 2, 3 ], [ 4, 5, 6 ] ]
gap> IsMutable(mat);
true
gap> MakeImmutable(mat[1]);
[ 1, 2, 3 ]
gap> IsMutable(mat);
true
gap> mat2 := TransposedMatDestructive(mat);
Error, Lists Assignment: <list> must be a mutable list in
...
```

**Comment by stevelinton** [comment]
@frankluebeck Yes. What I suggest is that if a matrix does not have all its rows mutable any *Destructive operation gives an error.  This should be in the documentation. IsLockedRepresentationVector only prevents Transposing in-place for a non-square matrix, which to me feels like such a technical condition that I'd delegate to TransposedMat and make and return a new matrix in that case, which the documentation allows.

**Comment by frankluebeck** [comment]
Wouldn't it be more convenient if the `...Destructive` methods would always just fall back to the non-destructive version when the destruction is not possible? Then one could write code which works independently of the mutability of the input and which is just more space efficient when possible.

I support unifying the documentation to "may destroy the input" and "may return an object identical to the input".

**Comment by stevelinton** [comment]
@frankluebeck  I'm fine with that. It's a slightly bigger change from existing documentation, but only in the direction of allowing more kids of arguments.
